### PR TITLE
[MB-16182] Auth added to update mto shipments endpoint

### PR DESCRIPTION
**Pull request by pearl-truss**
## Summary

This PR adds an authz check for the POST `/mto-shipments` endpoint. It checks that the service member associated with the session matches the service member associated with the move that the shipment will be created for

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [x] Have the Jira acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/run-storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

1. Start server and customer client.
2. Log in as a service member that has just submitted their move
3. Look at the network tab and find the value for the `MoveTaskOrderID`
4. Log out and log in as another service member to get their cookie info for postman. Load into postman based on however you set up postman.
    1. [Docs for setting up postman](https://transcom.github.io/mymove-docs/docs/tools/postman/setting-up-postman)
5. Make a patch request in postman to `/mto-shipments` 
    1. Since this is going to be a patcht request, you'll need to [set the `X-CSRF-TOKEN` header to be the `masked_gorilla_csrf` cookie value](https://transcom.github.io/mymove-docs/docs/tools/postman/intercepting-cookies-for-postman#form-a-working-get-and-patch-request-in-postman).
    2. You will also need a valid body to send along with the request. I used the following and note the `ShipmentID` was from the non logged in user (not the user I retrieved the cookies for)
```
{
    "estimatedWeight": 4200,
    "shipmentType": "HHG"
}
```
6. Verify you get a 404 not found response.
  
Screenshot 2023-07-26 at 11 19 11 AM

### CODE REVIEW

File: pkg/services/mto_shipment/mto_shipment_fetcher.go [hunk 1]
```diff
@@ -134,13 +134,11 @@ func (f mtoShipmentFetcher) ListMTOShipments(appCtx appcontext.AppContext, moveI
 func FindShipment(appCtx appcontext.AppContext, shipmentID uuid.UUID, eagerAssociations ...string) (*models.MTOShipment, error) {
 	var shipment models.MTOShipment
 	findShipmentQuery := appCtx.DB().Q().Scope(utilities.ExcludeDeletedScope())
-
```
Comment (pearl-truss) [inline pkg/services/mto_shipment/mto_shipment_fetcher.go:137]: I first tried variations of:
```
	if appCtx.Session() != nil && appCtx.Session().IsMilApp() {
     findShipmentQuery = findShipmentQuery.Where("mto_shipments.moves.orders.service_member_id =  ?" appCtx.Session().ServiceMemberId) 
	}
```

But I kept getting back a `QueryError` instead a `NotFound` error. @felipe-lee would you like me to spend more time to figure this out? I wasn't sure how much impact there was between adding the auth check in the query vs after the shipment was found

Comment (felipe-lee) [reply pkg/services/mto_shipment/mto_shipment_fetcher.go:137]: Sorry, I missed the notification for this till just now 😞 

Hmm that's weird that it gives back a query error. Does it have details on why?

The main impact is in terms of whether we end up loading data from the DB to the server memory that we don't end up needing.

So, if we can't do it with the query, it's not the worst outcome. It's more of a thing to optimize if we can.

**Review by felipe-lee — COMMENTED** [review]
Thanks for putting this together! I left a comment about a way we could do the query with the service member info, as well as some non-blocking suggestions and a question.

One thing that is blocking this for now is that we also need integration tests with middleware enabled. These live in the `pkg/handlers/routing` directory. Within that directory there are a packages for each of the APIs (internal, ghc, etc.). Could you add a file for the internal one for the `mto_shipments`? There are other examples in there, but here is the one I added for the upload amended orders authz fix I did: https://github.com/transcom/mymove/blob/main/pkg/handlers/routing/internalapi_test/orders_test.go

Log notes:
* Took me 1 hr, 15min
* Reviewed code changes in gh
* Pulled down code, started server and client, and then tried it out in postman as the instructions stated. It was then that I realized I mistyped the PR number when pulling it down, and I had to re-pull and restart everything, so I lost like 15-20 min here 😞 
* Most of my time was spent trying out different versions of the query and running with the debugger through all the pop code to try to figure out why the query that you tried using in https://github.com/transcom/mymove/pull/11098/files#r1272542034 wasn't working, because I thought it should work.

File: pkg/services/mto_shipment/mto_shipment_fetcher.go (hunk 1 above)
Comment (pearl-truss) [reply pkg/services/mto_shipment/mto_shipment_fetcher.go:137]: Thanks for looking into this more! One of the things I tried was using `InnerJoin` here and didn't think to try `LeftJoin`

File: pkg/services/mto_shipment/mto_shipment_fetcher.go [hunk 2]
```diff
@@ -150,6 +148,13 @@ func FindShipment(appCtx appcontext.AppContext, shipmentID uuid.UUID, eagerAssoc
 		}
 	}
 
+	if appCtx.Session() != nil && appCtx.Session().IsMilApp() {
```
[the remark this answers is not in the document]
Comment (pearl-truss) [reply pkg/services/mto_shipment/mto_shipment_fetcher.go:151]: I added this check because some of the test that use this function set context with `AppContextForTest()` which doesn't include session so those test end up with a nil pointer panic error

**Comment by pearl-truss** [comment]
> One thing that is blocking this for now is that we also need integration tests with middleware enabled. These live in the `pkg/handlers/routing` directory. Within that directory there are packages for each of the APIs (internal, ghc, etc.). Could you add a file for the internal one for the `mto_shipments`? There are other examples in there, but here is the one I added for the upload amended orders authz fix I did: https://github.com/transcom/mymove/blob/main/pkg/handlers/routing/internalapi_test/orders_test.go

Thanks Felipe! I didn't realize this was a part of this work and have been working on writing these test. I also want to call out that I didn't add this kind of testing to the POST mto shipments endpoint I worked on previously. Should that be flagged to be revisited in a new ticket?

File: pkg/services/mto_shipment/mto_shipment_fetcher.go (hunk 1 above)
Comment (felipe-lee) [reply pkg/services/mto_shipment/mto_shipment_fetcher.go:137]: Tbh, I always have to re-look up which join I need, and in this case didn't think about it and just got lucky. I followed the example in the pop docs for joining as I was testing and it just so happened to be the one that I needed for this query.

File: pkg/services/mto_shipment/mto_shipment_fetcher.go [hunk 3]
```diff
@@ -150,6 +148,13 @@ func FindShipment(appCtx appcontext.AppContext, shipmentID uuid.UUID, eagerAssoc
...
+
+		if shipment.MoveTaskOrder.Orders.ServiceMemberID != appCtx.Session().ServiceMemberID && shipment.MoveTaskOrder.Orders.ServiceMemberID != uuid.Nil {
```
[the remark this answers is not in the document]
Comment (felipe-lee) [reply pkg/services/mto_shipment/mto_shipment_fetcher.go:153]: Oh, interesting, good to know.

File: pkg/services/mto_shipment/mto_shipment_fetcher.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (felipe-lee) [reply pkg/services/mto_shipment/mto_shipment_fetcher.go:151]: Ah ok. Hmm, lmk if you disagree, but I think that's more of a test data setup issue that should be resolved there, rather than our code allowing for it. Like, there are things we do in the code that are for the sake of easier testing, like passing in dependencies (a.k.a dependency injection so we can mock them out in tests easily), or making functions do fewer things at a time to ease reasoning around what it does, but those also help with maintaining the code in terms of making easier to reason about things, and swap out dependencies if needed (e.g. if we change a 3rd party package that we use). But for this type of thing, it feels like tests that require a session should be setting up the app context with the session enabled. If it were a thing however, where we don't have a session in a regular flow, then I'd be fine with this. That's why I brought up the prime, I'm not sure if those requests have a session (I haven't checked), but if they don't, then it makes sense to keep this because I think some of our prime endpoints end up using this function.

**Comment by felipe-lee** [comment]
> > One thing that is blocking this for now is that we also need integration tests with middleware enabled. These live in the `pkg/handlers/routing` directory. Within that directory there are packages for each of the APIs (internal, ghc, etc.). Could you add a file for the internal one for the `mto_shipments`? There are other examples in there, but here is the one I added for the upload amended orders authz fix I did: https://github.com/transcom/mymove/blob/main/pkg/handlers/routing/internalapi_test/orders_test.go
> 
> Thanks Felipe! I didn't realize this was a part of this work and have been working on writing these test. I also want to call out that I didn't add this kind of testing to the POST mto shipments endpoint I worked on previously. Should that be flagged to be revisited in a new ticket?

Ah, yeah, we should add a ticket for that. Sorry, when we were trying to iron out what should be included in these tickets, the integration tests w/routing & middleware enabled were a last minute addition (literally came up as I worked on the first one), and I didn't do a good job of communicating that with the team. I need to go through the existing tickets to ensure they all have that as part of the AC, and say something about it in our team channel.

**Comment by pearl-truss** [comment]
> > > One thing that is blocking this for now is that we also need integration tests with middleware enabled. These live in the `pkg/handlers/routing` directory. Within that directory there are packages for each of the APIs (internal, ghc, etc.). Could you add a file for the internal one for the `mto_shipments`? There are other examples in there, but here is the one I added for the upload amended orders authz fix I did: https://github.com/transcom/mymove/blob/main/pkg/handlers/routing/internalapi_test/orders_test.go
> > 
> > 
> > Thanks Felipe! I didn't realize this was a part of this work and have been working on writing these test. I also want to call out that I didn't add this kind of testing to the POST mto shipments endpoint I worked on previously. Should that be flagged to be revisited in a new ticket?
> 
> Ah, yeah, we should add a ticket for that. Sorry, when we were trying to iron out what should be included in these tickets, the integration tests w/routing & middleware enabled were a last minute addition (literally came up as I worked on the first one), and I didn't do a good job of communicating that with the team. I need to go through the existing tickets to ensure they all have that as part of the AC, and say something about it in our team channel.

No worries! I've been looking at the test you added for amended orders as a reference for a patch request. And I'm honestly a bit stuck on this. I'm getting back a 422 (previously I was getting a 415 until I changed to setting the content type for the header to be `application/json` 

```
	setUpRequestBody := func() *bytes.Buffer {
		f := ShipmentUpdate{
			EstimatedWeight: "5000",
		}

		var buf bytes.Buffer
		err := json.NewEncoder(&buf).Encode(f)
		if err != nil {
			log.Fatal(err)
		}

		return &buf
	}
	suite.Run("Unauthorized mto shipment update by another service member", func() {
		shipment := factory.BuildMTOShipment(suite.DB(), nil, nil)

		maliciousUser := factory.BuildServiceMember(suite.DB(), factory.GetTraitActiveServiceMemberUser(), nil)

		endpointPath := fmt.Sprintf("/internal/mto-shipments/%s", shipment.ID.String())
		
		body := setUpRequestBody()

		req := suite.NewAuthenticatedMilRequest("PATCH", endpointPath, body, maliciousUser)

		req.Header.Set("Content-Type", "application/json")

		rr := httptest.NewRecorder()

		suite.SetupSiteHandler().ServeHTTP(rr, req)

		suite.Equal(http.StatusNotFound, rr.Code)
	})
```

File: pkg/services/mto_shipment/mto_shipment_fetcher.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pearl-truss) [reply pkg/services/mto_shipment/mto_shipment_fetcher.go]: @felipe-lee I started implementing this change. I updated multiple test across about 10 files but now I'm seeing that without this check it's failing for a test in the prime api handler for deleting a shipment. So I think your first comment was correct that we need this since it's called by the prime and the prime doesn't have a session

File: pkg/handlers/routing/internalapi_test/mto_shipments_test.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		req.Header.Set("If-Match", etag.GenerateEtag(shipment.UpdatedAt))
+
+		rr := httptest.NewRecorder()
+		suite.SetupSiteHandler().ServeHTTP(rr, req)
+		resBody, err := io.ReadAll(rr.Body)
+		if err != nil {
+			fmt.Printf("client: could not read response body: %s\n", err)
+		}
+
+		fmt.Printf("resbody: %s\n", resBody)
```
Comment (felipe-lee) [inline pkg/handlers/routing/internalapi_test/mto_shipments_test.go:43]: **suggestion (non-blocking)**: could you remove this bit, since we mainly used it for debugging? Or if you think it'd be useful to keep, then maybe some assertion can be made on the body, like it having the "Not Found" text or something.

File: pkg/services/mto_shipment/mto_shipment_fetcher.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (felipe-lee) [reply pkg/services/mto_shipment/mto_shipment_fetcher.go]: Ah, ok. What do you think is the best way to communicate this to folks so that others (or future us if we forget 😅) know that we need this? Like do you think adding a comment here would be good? Or maybe more of a docs thing that services in general need to check for session before using it if they may be used by the prime? 🤔

**Review by felipe-lee — APPROVED** [review]
I have one blocking comment, but otherwise this lgtm and I'll approve so that you aren't truly blocked by me.

File: pkg/services/mto_shipment/mto_shipment_updater.go [hunk 5]
```diff
@@ -297,6 +297,7 @@ func (e StaleIdentifierError) Error() string {
 // UpdateMTOShipment updates the mto shipment
 func (f *mtoShipmentUpdater) UpdateMTOShipment(appCtx appcontext.AppContext, mtoShipment *models.MTOShipment, eTag string) (*models.MTOShipment, error) {
 	eagerAssociations := []string{"MoveTaskOrder",
+		"MoveTaskOrder.Orders",
```
[the remark this answers is not in the document]
Comment (pearl-truss) [reply pkg/services/mto_shipment/mto_shipment_updater.go]: @felipe-lee sorry I missed that comment previously. I checked and it's not necessary so I removed it

**Comment by pearl-truss** [comment]
@felipe-lee Agreed it should be documented somewhere. I'm thinking [this](https://transcom.github.io/mymove-docs/docs/backend/guides/use-stateless-services-with-app-context) could be a good place to add that documentation?

**Comment by felipe-lee** [comment]
> @felipe-lee Agreed it should be documented somewhere. I'm thinking [this](https://transcom.github.io/mymove-docs/docs/backend/guides/use-stateless-services-with-app-context) could be a good place to add that documentation?

Sounds good, do you mind creating a ticket for that?
